### ISSUE TITLE
possibility of "dark option"

### ISSUE THREAD

**Issue by uklomp** [Signbank/Global-signbank#1166]
We'd like Global Signbank to have an optional "dark mode", to improve accessibility for people with low vision / deafblind people. Would this be possible? It does not have priority. 

An example can be found at the Oogvereniging website, where the dark mode is black and white (Visible if you click on 'Aa')

**Comment by susanodd** [comment]
Yes, this would be possible. Although we may need to update Bootstrap. (Need to investigate. There is a dark mode setting in the Admin.)

**Comment by susanodd** [comment]
With minimal effort and a bit of css settings modification to the baselayout and menu templates, this is possible.
Shown here with a "dark mode" button in the header. A setting could be added in the user profile. Or it could be some kind of "view assistance" on/off thing.
The layout is a bit weird looking since I didn't modify the bootstrap "visual layout structure" things. (Like the things you click on.)

![toggle-dark-mode-button-header](https://github.com/Signbank/Global-signbank/assets/24272544/781cc2f9-2c9c-4ad2-a71e-f96125d7a7de)

![dark-mode-toggle-view](https://github.com/Signbank/Global-signbank/assets/24272544/418a3d06-8e2d-4226-8f74-40608902fb53)

![toggle-dark-mode-buttons-open-toggle-view](https://github.com/Signbank/Global-signbank/assets/24272544/54766b2f-17be-4f3b-880a-9b818c140689)

**Comment by susanodd** [comment]
I made the names on the actual buttons black black so they pop a bit more.

@uklomp can you perhaps get feedback on "what needs improvement" still ?

This is not a real dark mode, since it's a bit of a hack.
To do this elegantly, we need to upgrade Bootstrap.

**Comment by Woseseltops** [comment]
To be totally clear, @uklomp , does it only need to be dark mode, or also high contrast mode? If only dark, I think the dark mode of (for example) Github is pretty, so we could use it for inspiration.

**Comment by uklomp** [comment]
The request we received was for a dark mode. We also received a request for more contrast on the "normal" website, I still need to make an issue for that. 

@susanodd  I'm happy to ask these people for feedback on the dark mode. It's easiest if you give me a link or certain screenshots. Do you want me to do this now already, with the above screenshots you provided?

**Comment by susanodd** [comment]
@uklomp this is on signbank-susan now.

https://signbank-susan.cls.ru.nl/

If you have multiple different signbanks open, you may need to close them all and open it in a new browser window.

(So shut the browser and logout and login to signbank if dark mode doesn't work when you first try it. Otherwise it's fine.)

There is a button on each page, in the header. So you need to click it again to turn it on again.

It's totally ugly looking at the moment, apologies :)

Some feedback about "how black" or "how white" or "how high contrast" would be nice.

The buttons, etc still need to be done.

**Comment by susanodd** [comment]
Gloss Detail View hasn't been done yet. (It's still white in Dark Mode.)

**Comment by uklomp** [comment]
Hi Susan,
Would it be possible to show the dark mode without the necessity to log in? If not, I'll create guest accounts for the two feedback givers.

**Comment by susanodd** [comment]
Oh, you're right, I see it's not on the site if not logged in.
I'll have to see why that is.
It's probably best to create guest accounts and view access for NGT (that should be automatic from the site).

I'll try to add the button for anonymous users. Although I don't know off the top of my head why it's not there.

**Comment by uklomp** [comment]
OK, I'll email them with these questions.

**Comment by uklomp** [comment]
Great! Thank you.

**Comment by susanodd** [comment]
@uklomp dark mode is available now without logging in.

**Comment by uklomp** [comment]
I emailed our feedback persons. Will get back to you once they replied.

**Comment by susanodd** [comment]
![dark-mode-twilight-style](https://github.com/Signbank/Global-signbank/assets/24272544/2e6fc3c2-b289-498e-a2e0-bd067745d342)

**Comment by susanodd** [comment]
![dark-mode-twilight-search-results](https://github.com/Signbank/Global-signbank/assets/24272544/9223a43c-b968-4681-b4a6-51e0050331ca)

**Comment by susanodd** [comment]
It really is easier on the eyes if you compare the pinky text versus the whiter search results. The white looks really harsh at night.

**Comment by uklomp** [comment]
I receveived a few feedback points: 
1. The greyish whitish text is fine for now, but when the text is bold, it should also be very clear bold and white (apparently it wasnt perceived as such)
The user recommended this website for more examples https://blog.felgo.com/qt-dark-mode-with-ios-and-android-native-components

2. Some panels and menus are grey, and their boundary lines are white. In regular mode, there is hardly a difference between the color of the panels and their lines, but in dark mode, the lines are suddenly very visible. It is better to make this difference smaller.

3. When in dark mode, the dark mode option is still called dark mode. Maybe make sort of on/off button or dark/light button which changes along. Some suggestions:  https://linku.nl/dark-mode/

**Comment by susanodd** [comment]
I revised the code a lot in the weekend to make it reddish tones instead of blue. (@Woseseltops suggested to make it match the signbank banner.)

I made css-variables for most (3/4) of the colours so it will be easier to adjust to simply change the colours.

(The colours code is not finished yet, since because of the "surprise" things that appear in dark mode, then some more inheritance needs to be added with yet more overrides of Bootstrap....)

The actual "lines" in the pages are a huge headache. (This seems to be hard-coded into Bootstrap that things "have" borders on them. I haven't figured out how to get rid of some of them that all of a sudden appear when in dark mode.)

**Comment by susanodd** [comment]
@uklomp I use dark mode a lot on my iPhone. Mostly Health apps.

But one thing they all share is the use of bright colours for the semantics of the different things you click on.

So if applied to Signbank, then we end up with a rainbow of colours. (This becomes very obvious with the buttons. Signbank light mode just uses light grey for all the non-main buttons. But this looks really bad in dark mode, since they distract from the rest of the page if too many things are white or light grey. Hence the use of colours in Health apps, to help find the things you're looking for.)

I made a start on the buttons for the "page" that is selected in the list. I made it kind of pink. (It was a really shadowy grey, as it is on light mode signbank.)

**Comment by susanodd** [comment]
![dark-mode-hits-per-page-button-pink](https://github.com/Signbank/Global-signbank/assets/24272544/44cb38cd-8cd1-4587-a376-65585befe0c8)

**Comment by susanodd** [comment]
If this was Apple Health, each of the different categories of Search By would be a different colour. On a tiny mobile device that is kind of necessary.

Above, anything white stands out a bunch. But the "blue" default buttons also stand out.
On GitHub in dark mode, they use Green and Blue for button type things. (I need to ask @Woseseltops about using different colours. I was thinking maybe the banner needs to be something other than red so we can use meaningful colours to identify things.)

**Comment by susanodd** [comment]
NOTE TO DEVELOPERS: I removed the "Add New Gloss" code that exists right in the middle of the Search Panel. (You see the button in the screenshot above.) This is impossible to format inside of this template because it needs extension "exceptions to the rule" in the html markup structure. Moreover there is a nasty lookahead for either "choosing an exiting lemma" or "creating a new lemma" in the middle of this code. So I gave up formatting in for dark mode and just removed it. Users may create new glosses via the Menu.)

**Comment by susanodd** [comment]
@uklomp I put the latest version of the dark mode on signbank-susan

Those red colours do look a bit frightful. (On Safari at home in the evening, they look rather nice. But not here on Firefox in fluorescent light.)

I wear blue light filter glasses 100% of the time at the computer. So that may affect how I perceive the colours.

But it does look way too red now.

**Comment by uklomp** [comment]
I agree that it's too red now. I can't judge if we can already send to the feedback givers already or that you want to change it first again ? I didn't really understand what was wrong with grey/white?

**Comment by susanodd** [comment]
I found a link that explains what dark mode should accomplish

https://blog.zeplin.io/dark-mode-color-palette

> The purpose of Dark Mode is to make screens easier to read. A direct color inversion often has the opposite effect, since colors that contrast nicely in a light setting don't necessarily have complements that pair the same way.

That is the problem I'm having.

It doesn't work very well to follow the Bootstrap element structure.
That's also altered by Django template code.

**Comment by susanodd** [comment]
@Woseseltops when you're reviewing the pull request that's on signbank-dev, see how I've done the new affiliation tags.

I was able to put the color that's defined in the Admin for the affiliations into the tags.

So I would assume we could somehow follow this to make it possible to set the colors for dark mode?

Most of them are in variables at the top of the style.css file.

The other new css files are basically to toggle it, and to override zillions of hard-coded colors, like for select2.

**Comment by uklomp** [comment]
> I found a link that explains what dark mode should accomplish
> 
> https://blog.zeplin.io/dark-mode-color-palette
> 
> > The purpose of Dark Mode is to make screens easier to read. A direct color inversion often has the opposite effect, since colors that contrast nicely in a light setting don't necessarily have complements that pair the same way.
> 
> That is the problem I'm having.
> 
> It doesn't work very well to follow the Bootstrap element structure. That's also altered by Django template code.
> 
> You can't make shadow boxes on really dark things because it's just more black. That's why I've made the border lines visible to distinguish where things are.
> 
> Anybody have other ideas?
> 
> Using GitHub as an example doesn't work. No hordes of context-sensitive dynamically generated pull-down menus on GitHub.
> 
> See the nice blue outline on stuff in GitHub? What colour should this be in Sginbank Dark Mode ?

Where can I find the dark mode on GitHub? 
> 
> I've used blue the same as GitHub but it looks awful with the Signbank banner. Now it's using reddish tints. But then you need reddish shades in the boxes, otherwise it looks really ugly. Now it's all become Twilight saga dark mode.

**Comment by susanodd** [comment]
I think GitHub uses your browser. GitHub dark mode is turned on automatically on my Apple because I have it turned on for the computer.
It might be under Settings for your account?

**Comment by uklomp** [comment]
Where can I find dark mode again? I can't find it on signbank-susan.

**Comment by susanodd** [comment]
It's not on any server at the moment, because the four development servers are being used for other pull requests.
I'll put it back when one comes free.

**Comment by susanodd** [comment]
@uklomp this is on signbank-susan again.

I revised it to include the recent changes.

I removed all the red hues from the dark parts.

There are a few pages that still need work but much of it is filled in now.

**Comment by susanodd** [comment]
@Jetske question about the Senses modals in Gloss Edit.

Is there a reason why you hand-coded this instead of using the Bootstrap modals?

The style inherit isn't working on the senses modals. I'm going to replace the background and text colors with a variable.

**Comment by Jetske** [comment]
@susanodd probably because I don't know how to do it with Bootstrap, but feel free to make changes

**Comment by uklomp** [comment]
I requested feedback from two signers with vision impairments, to me it looks great though

**Comment by Woseseltops** [comment]
Let's close this for now and open separate more specific issue if vision impaired users have special users!
